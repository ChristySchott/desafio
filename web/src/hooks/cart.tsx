import React, {
  createContext,
  useState,
  useCallback,
  useContext,
  useEffect,
} from 'react';

export interface ProductInterface {
  id: number;
  sku: string;
  path: string;
  name: string;
  image: string;
  price: number;
  specialPrice?: number;
  quantity: number;
}

interface CartContext {
  products: ProductInterface[];
  addToCart(item: Omit<ProductInterface, 'quantity'>): void;
  increment(id: number): void;
  decrement(id: number): void;
}

const CartContext = createContext<CartContext | null>(null);

const CartProvider: React.FC = ({ children }) => {
  const [products, setProducts] = useState<ProductInterface[]>([]);

  useEffect(() => {
    async function loadProducts(): Promise<void> {
      const storagedProducts = localStorage.getItem('@Webjump:products');

      if (storagedProducts) {
        setProducts([...JSON.parse(storagedProducts)]);
      }
    }

    loadProducts();
  }, []);

  const addToCart = useCallback(
    async product => {
      const productExists = products.find(item => item.id === product.id);

      if (productExists) {
        setProducts(
          products.map(p =>
            p.id === product.id ? { ...product, quantity: p.quantity + 1 } : p,
          ),
        );
      } else {
        setProducts([...products, { ...product, quantity: 1 }]);
      }

      localStorage.setItem('@Webjump:products', JSON.stringify(products));
    },
    [products],
  );

  const increment = useCallback(
    async id => {
      const updatedProducts = products.map(product =>
        product.id === id
          ? { ...product, quantity: product.quantity + 1 }
          : product,
      );

      setProducts(updatedProducts);

      localStorage.setItem(
        '@Webjump:products',
        JSON.stringify(updatedProducts),
      );
    },
    [products],
  );

  const decrement = useCallback(
    async id => {
      const updatedProducts = products.map(product =>
        product.id === id
          ? { ...product, quantity: product.quantity - 1 }
          : product,
      );

      setProducts(updatedProducts);

      localStorage.setItem(
        '@Webjump:products',
        JSON.stringify(updatedProducts),
      );
    },
    [products],
  );

  const value = React.useMemo(
    () => ({ addToCart, increment, decrement, products }),
    [products, addToCart, increment, decrement],
  );

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
};

function useCart(): CartContext {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error(`useCart must be used within a CartProvider`);
  }

  return context;
}

export { CartProvider, useCart };
