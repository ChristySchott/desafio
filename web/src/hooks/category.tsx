import React, { createContext, useContext, useState } from 'react';

interface CategoryContextData {
  category: string;
  setCategory: React.Dispatch<React.SetStateAction<string>>;
}

const CategoryContext = createContext<CategoryContextData>(
  {} as CategoryContextData,
);

const CategoryProvider: React.FC = ({ children }) => {
  const [category, setCategory] = useState('calcados');

  return (
    <CategoryContext.Provider
      value={{
        category,
        setCategory,
      }}
    >
      {children}
    </CategoryContext.Provider>
  );
};

function useCategory(): CategoryContextData {
  const context = useContext(CategoryContext);

  if (!context) {
    throw new Error('useCategory must be used within an CategoryContext');
  }

  return context;
}

export { CategoryProvider, useCategory };
