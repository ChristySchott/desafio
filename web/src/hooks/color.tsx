import React, { createContext, useContext, useState } from 'react';

interface ColorContextData {
  colorToFilter: string;
  setColorToFilter: React.Dispatch<React.SetStateAction<string>>;
}

const ColorContext = createContext<ColorContextData>({} as ColorContextData);

const ColorProvider: React.FC = ({ children }) => {
  const [colorToFilter, setColorToFilter] = useState('');

  return (
    <ColorContext.Provider
      value={{
        colorToFilter,
        setColorToFilter,
      }}
    >
      {children}
    </ColorContext.Provider>
  );
};

function useColorFilter(): ColorContextData {
  const context = useContext(ColorContext);

  if (!context) {
    throw new Error('useColorFilter must be used within an ColorContext');
  }

  return context;
}

export { ColorProvider, useColorFilter };
