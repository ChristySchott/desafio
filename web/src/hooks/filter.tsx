import React, { createContext, useContext, useState } from 'react';

interface FilterContextData {
  filter: string;
  setFilter: React.Dispatch<React.SetStateAction<string>>;
}

const FilterContext = createContext<FilterContextData>({} as FilterContextData);

const FilterProvider: React.FC = ({ children }) => {
  const [filter, setFilter] = useState('');

  return (
    <FilterContext.Provider
      value={{
        filter,
        setFilter,
      }}
    >
      {children}
    </FilterContext.Provider>
  );
};

function useFilter(): FilterContextData {
  const context = useContext(FilterContext);

  if (!context) {
    throw new Error('useFilter must be used within an FilterContext');
  }

  return context;
}

export { FilterProvider, useFilter };
