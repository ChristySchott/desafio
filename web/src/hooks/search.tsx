import React, { createContext, useContext, useState } from 'react';

interface SearchContextData {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

const SearchContext = createContext<SearchContextData>({} as SearchContextData);

const SearchProvider: React.FC = ({ children }) => {
  const [search, setSearch] = useState('');

  return (
    <SearchContext.Provider
      value={{
        search,
        setSearch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

function useSearch(): SearchContextData {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error('useSearch must be used within an SearchContext');
  }

  return context;
}

export { SearchProvider, useSearch };
