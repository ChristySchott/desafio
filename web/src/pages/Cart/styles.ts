import styled from 'styled-components';

export const Container = styled.div`
  max-width: 1048px;
  margin: 0 auto;
  position: relative;

  ul {
    list-style: none;
    width: 100%;
  }
`;
