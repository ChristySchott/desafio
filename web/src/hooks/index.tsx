import React from 'react';

import { FilterProvider } from './filter';
import { ColorProvider } from './color';
import { CartProvider } from './cart';

const AppProvider: React.FC = ({ children }) => {
  return (
    <FilterProvider>
      <ColorProvider>
        <CartProvider>{children}</CartProvider>
      </ColorProvider>
    </FilterProvider>
  );
};

export default AppProvider;
