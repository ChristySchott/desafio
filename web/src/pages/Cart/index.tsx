import React from 'react';

import CartProducts from 'components/CartProducts';
import { Container } from './styles';

const Cart: React.FC = () => {
  return (
    <Container>
      <ul>
        <CartProducts />
      </ul>
    </Container>
  );
};

export default Cart;
