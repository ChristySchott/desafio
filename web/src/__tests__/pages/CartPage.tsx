import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

import { CartProvider } from 'hooks/cart';
import Cart from '../../pages/Cart';

describe('Cart Page', () => {
  it('should render with no error', () => {
    const { container } = render(
      <BrowserRouter>
        <CartProvider>
          <Cart />
        </CartProvider>
      </BrowserRouter>,
    );

    expect(container).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
