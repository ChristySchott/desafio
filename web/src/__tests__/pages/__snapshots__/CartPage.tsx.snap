// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Page should render with no error 1`] = `
<div
  class="sc-fujyUd ksfBge"
>
  <ul>
    <div
      class="sc-bdnylx jHJelA"
    >
      <header>
        <h2>
          Carrinho
        </h2>
      </header>
      <ul
        class="sc-gtssRu eWCgwA"
      />
      <div
        class="sc-dlnjPT lcvNAZ"
      >
        <svg
          color="#c61a28"
          fill="none"
          height="50"
          size="50"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="color: rgb(198, 26, 40);"
          viewBox="0 0 24 24"
          width="50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="9"
            cy="21"
            r="1"
          />
          <circle
            cx="20"
            cy="21"
            r="1"
          />
          <path
            d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
          />
        </svg>
        <p>
          O carrinho est√° vazio.
        </p>
      </div>
    </div>
  </ul>
</div>
`;
